### ISSUE TITLE
2.18 - (chip tbd) - FLUXX3

### ISSUE THREAD

**Issue by reed-dan** [RoboDurden/Hoverboard-Firmware-Hack-Gen2.x#52]
Brand: Fluxx (GoTrax)
Model: Fluxx3
URL: https://gotrax.com/products/fluxx-fx3-self-balancing-hoverboard
Manufacturer Date: 02/05/2020
Max: Speed 6mph Weight 220lb 
Battery: 25.2 V 2.6Ah
Charger: Barrel Jack 24V 0.6A

Body: Aluminium
Wheels: Plastic 6.5in
Board Configuration: 2 boards
Chip: TBD (130)
Notes: Battery connector is XT30

Similar to [2.17](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/issues/44 )

**Comment by AILIFE4798** [comment]
2.18 is used so it would be 2.19
it should be pin compatiable with 2.17 test with constant current supply

**Comment by RoboDurden** [comment]
Yes this will be 2.19
Layout is quite different from 2.17 (also GoTrax) so it will need some pin tracing and testing with constant current supply (cheapest way: LM2596HVS 60V 3A Buck Converter = https://www.aliexpress.com/item/4001360810712.html )

@reed-dan please upload nice top view photos from master and slave and master-backside and slave-backside, this will make pin tracing easier.
A description of the pin headers would also be nice.
I guess the one with the buzzer in the middle is the master as it has the onOff button and the charger socket.

**Comment by RoboDurden** [comment]
Okay i have added this as 2.19 to the repo.
@reed-dan the 6 mosfet are compatible to my new autodetect firmware:
![hall pins](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/82aa863c-c531-49de-8d5e-19c891d46cb5)

But i can not really see where the two SWD pins for flashing with st-link dongle goes:

PA14: SWCLK
PA13: SWDIO

And you need a uart communicaton for cheap usb-uart-dongle (or esp32 with my autodetect.ino)

PA2/PA3 is usually the  master/slave header:
![uart1 maybe](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/868ba5dd-7adc-4d3a-94b0-ecd238df02b9)

PB6 and PB7 could be on the empty header to the left:
![uart0 maybe](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/6b9d6970-24d3-49c8-b5b2-15b9911d98fe)
But then PB6 and PB7 seem to be bridged. So you would have to access the still seperate PB6 and PB6 before the final smd resistors.

Best start with my hoverboard autodetect tx=PA2, rx=PA3.bin at https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/tree/main/BinariesToTest

**Comment by AILIFE4798** [comment]
there really is no swd connector
thats gonna be a bit hard to solder wire to pin of mcu but it can be done

**Comment by AILIFE4798** [comment]
i guess the swd is disabled anyways just use serial to flash

**Comment by RoboDurden** [comment]
please upload some screentshots on how to flash via serail and not swd.
To me it looks like PA13 and pA14 at least a connected to some smd resistors/capacitors. So soldering thin cable to those should not be such a big problem.

**Comment by AILIFE4798** [comment]
but the problem is if the pin used for something else then swd have to be disabled anyways

**Comment by AILIFE4798** [comment]
well the bootloader recover uart pin is a9 a10 which is used by motor so it cannot be used also
thats gonna be a bit tricky

**Comment by AILIFE4798** [comment]
![image](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/142502122/b0424531-bb99-471f-9b07-fd2c616b64e1)

**Comment by AILIFE4798** [comment]
pattern1=boot0 pullup boot1 pulldown

**Comment by RoboDurden** [comment]
@AILIFE4798  please wait for @reed-dan to start trying. We already have a board that also uses a SWD pin for something LED if i remember correctly. In this case we only have to take care to not use this pin as something like LED_RED because after that it needs the NREST pin to be able to flash again.

**Comment by reed-dan** [comment]
> But i can not really see where the two SWD pins for flashing with st-link dongle goes:
> 
> PA14: SWCLK PA13: SWDIO
> 
The swclk swdio pins are filled in with solder but to the left of the mcu

![Screenshot_20240118_164213_Gallery](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/0a7774c4-29be-479e-aeb5-6bdd5eb76093)

I will get some better pictures soon. Sorry been wokring  on 2.8.

**Comment by reed-dan** [comment]
![20240113_142535](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/aa7c108f-55f5-458d-ad26-aa831aed5e30)
![20240118_174257](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/2c76b2a4-375c-4e4a-9610-116eca603c07)
![20240118_174314](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/20179705-763d-4d21-849c-aee57368c848)
![20240118_174533](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/932c78aa-854c-41be-8258-75d1c0d149de)
![20240118_174543](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/eadd2ea3-766e-4d94-b38e-4eff68def422)

**Comment by reed-dan** [comment]
![20240118_174257](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/3b894887-676c-4d0f-8251-595096f89436)
![20240118_174314](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/eb10b9d3-9c91-463d-970b-cc9b3e07b889)
![20240118_174533](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/20217356-c45d-4dd8-aad7-2d13608048eb)
![20240118_174543](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/c71c055f-7036-46c9-9b47-937770742706)

**Comment by reed-dan** [comment]
How do you determine which board is the Slave & Master?

**Comment by reed-dan** [comment]
Tracing Work in Progress
MCU
Pin 30-PA9->R83->EG2133 Pin 2
Pin 28-PB15->R59->EG2133 Pin 4
Pin 26-PB13->R82->EG2133 Pin 6
Pin 31-PA10->R60->EG2133 Pin 1
Pin 29-PA8->R85->EG2133 Pin 3
Pin 27-PB14->R84->EG2133 Pin 5

Pin 37-PA14: SWCLK
Pin 34-PA13: SWDIO

Mosfet
Q2: D:R38
Q8: D:R39

**Comment by reed-dan** [comment]
![20240118_212222](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/1dd731e1-97c3-4161-a40b-c769e759fd8b)

**Comment by AILIFE4798** [comment]
The board with power button is master
Most of the time they write it on the board with Chinese but apparently this time not
You should not trace the pin should wait for auto detect
Only trace tx rx and the swd pins

**Comment by RoboDurden** [comment]
Yes . Pb6 and pb7 would be good if you plan to use the master slave cable.

**Comment by RoboDurden** [comment]
I am still in bed.
If you have Keil installed you can compile the latest autodetect yourself.
Your mosfet pins are compatible.
And the hall sensors detection worked for ailife.

**Comment by AILIFE4798** [comment]
is the serial fixed now

**Comment by reed-dan** [comment]
It will be a couple of days, I am waiting on desoldering wick so I can take the solder out of the swd holes. It looks like PB6 & PB7 are both unused on the board. Would this something I would need to have connection to permanently or just temp? If it is just temp I ordered some micro probe clips and may be able to hook onto the pins, otherwise I am not sure I can solder to those pins at this time.

Good news is I have a 2.9 board with an gd32f103c8t6 (instead of  gd32f103c6t6) chip that has an "external interface" appears to be connected to PB6 & PB7!

**Comment by AILIFE4798** [comment]
i dont think any micro probe can hold onto a qfn48 without shorting out lol
soldering copper wire that is used in motor winding is the easiest way if it is really the case that is unused
but pb6 and pb7 is the only hardware i2c pin so if they want to use hardware i2c for IMU it would be much easier thrn soldering to gd32's leg

**Comment by reed-dan** [comment]
I probed them to PB5 pin 41 (Alternate: SPI0_MOSI, I2C0_SMBA, TIMER15_BRKIN,
TIMER2_CH1) does that seem correct?

![tempFileForShare_20240119-110457](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/50810303-c0ba-4839-a814-0435c5f77def)

**Comment by RoboDurden** [comment]
PB5 can not be used as serial0.

Why do i not see the back connections on your transparent pcb photo ?
![grafik](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/3cb1dad8-c394-440d-8177-e32676e21ba2)
![grafik](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/1b9287d9-8916-40f6-99b3-0f76afa7953e)

Am i mixing wrong layouts ?

**Comment by reed-dan** [comment]
I didn't upload the back side transparent, but you are spot on. I tested those pads previously with my fine point meter probes, but they seemed to no be connected to anything. When I got out the needles we are making progress, and there are nice pads to solder to.

![Screenshot_20240119_010544_Gallery](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/36dd8769-8b91-4212-9c95-dc838ee8d9ac)
![tempFileForShare_20240119-122155](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/60077676/1f9b419c-4434-4f1e-bc6a-ef3df5d2ea31)

**Comment by RoboDurden** [comment]
But you can see the footprint of a bluetooth module there :-)
![grafik](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/000fdf29-7f46-43dd-a4f8-dfa98758be5d)

**Comment by RoboDurden** [comment]
![Uploading uart0.jpg…]()

**Comment by RoboDurden** [comment]
![uart0](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/assets/12238841/16430fe5-2592-497e-8507-1caafb0408d2)

**Comment by AILIFE4798** [comment]
to easily solder pin header without removing the solder in the hole first you can first cut the pin header into a single pin and solder it 1 by 1(melt the solder from one side of board and push the pin header in)

**Comment by reed-dan** [comment]
I desoldered the swd header pin holes today, but the holes are smaller than standard header pins.

**Comment by RoboDurden** [comment]
https://youtu.be/qVJqOmmUm2c

**Comment by danil212211** [comment]
I have this layout. I resoldered those pins and connected to it. Unfortunately I couldn't use autodetect binary from this repo, cause it's 33kb and GD32F130C6T6 has only 32kb of flash memory. 

I built autodetect from sources, just changed GCC optimization flag to -Oz, and it flashed alright. But sadly, when I tried to detect hall sensors, it didn't find anything. For power I used my lab supply with 24v 1a.

**Comment by RoboDurden** [comment]
> GCC optimization flag to -Oz, 

With platformio ?
Please post screenshot.

User interface of my autodetect running ?
Can you detect led pins ?

**Comment by danil212211** [comment]
> With platformio ?

I used Keil uVision, set Autodetect macros on in config.h

>  Please post screenshot.
I will post screenshot in an hour or two, cant do it now.

> User interface of my autodetect running ? Can you detect led pins ?

Yes, I used your autodetect interface, I haven't tried detecting led pins, but successfully detected buzzer pin on C13

**Comment by RoboDurden** [comment]
You could make bHall a global variable and monitor it (together with iTest) with StmStudio (or MCUViewer):

**Comment by danil212211** [comment]
> You could make bHall a global variable and monitor it (together with iTest) with StmStudio (or MCUViewer):
> 
> ` uint8_t bHall = digitalRead(aoPin[iTest].i); `
> 
> If you already see the pin currently tested in my user interface (via uart), only visualizing bHall might help you to identify the hall sensors. Of course, the motor should be spinning.
> 
> my hall identifier was tuned manually:
> 
> ` if ( (iTime > 5) && (iTime < 50) ) // the hall on-time should match the rotation speed `
> 
> Make `uint16_t iTime = msTicks-msTicksOld;` also a global variable to see why my tuning does not work for your motor.
> 
> Read here in the wiki for StmStudio.

Ok, thank you! Maybe threre is some problems with my motors, cause they haven't rotated a bit while test was running, Ill try debugging with your approach and will post results if Ill find anything

**Comment by danil212211** [comment]
Yeah, I traced the pins to EG2133 and looked at the datasheet. It seems that I have a faulty board, and I haven't checked if it works before starting disassembling it.

**Comment by RoboDurden** [comment]
# Tell me, why didn't you post a screenshot with this `GCC optimization flag to -Oz,` ?

**Comment by RoboDurden** [comment]
Awesome, now even RTT_REMOTE is below 32kB:

```
REMOTE_AUTODETECT with Optimization "O1"

REMOTE_USART
Program Size: Code=32028 RO-data=1584 RW-data=648 ZI-data=2112  

RTT_REMOTE
Program Size: Code=35948 RO-data=1628 RW-data=648 ZI-data=3280  


REMOTE_AUTODETECT with Optimization "Oz Image Size"


RTT_REMOTE + WINDOWS_RN
Program Size: Code=24604 RO-data=1596 RW-data=644 ZI-data=3268  

REMOTE_USART
Program Size: Code=23912 RO-data=1572 RW-data=644 ZI-data=2100  
```


> Yeah, I traced the pins to EG2133 and looked at the datasheet and there is no enable pins, and I haven't checked if it works before starting disassembling it, so It seems that I have a faulty board
UPD: The second board has a short circuit, so It's 100% faulty=(

Without analog mosfet driver, it should be possible to repair such a board.
Short cut is probably because of two mofets (lo and hi side of one phase) have burned.
Remove them and see with an oscilloscope if the EG2133 is still outputting the gate voltages for the broken mosfests.
If not, you would have to replace the EG2133 as well.

**Comment by danil212211** [comment]
> Awesome, now even RTT_REMOTE is below 32kB:

Glad I could help=)

> Without analog mosfet driver, it should be possible to repair such a board. Short cut is probably because of two mofets (lo and hi side of one phase) have burned. Remove them and see with an oscilloscope if the EG2133 is still outputting the gate voltages for the broken mosfests. If not, you would have to replace the EG2133 as well.

Got it. I think it is too much trouble just for making motor spin=) I think I will use bldc driver for my project instead, would be a lot easier
